packets.go

```diff
@@ -305,16 +310,37 @@ func (mc *mysqlConn) writeHandshakeResponsePacket(authResp []byte, plugin string
 	var authRespLEIBuf [9]byte
 	authRespLen := len(authResp)
 	authRespLEI := appendLengthEncodedInteger(authRespLEIBuf[:0], uint64(authRespLen))
-	if len(authRespLEI) > 1 {
+	if len(authRespLEI) > 1 && clientFlags&clientSecureConn != 0 {
 		// if the length can not be written in 1 byte, it must be written as a
 		// length encoded integer
 		clientFlags |= clientPluginAuthLenEncClientData
 	}
 
 	pktLen := 4 + 4 + 1 + 23 + len(mc.cfg.User) + 1 + len(authRespLEI) + len(authResp) + 21 + 1
+	if clientFlags&clientSecureConn == 0 || clientFlags&clientPluginAuthLenEncClientData == 0 {
+		pktLen++
+	}
+
+	connectAttrsBuf := make([]byte, 0, 100)
+	if mc.flags&clientConnectAttrs != 0 {
+		clientFlags |= clientConnectAttrs
+		connectAttrsBuf = appendLengthEncodedString(connectAttrsBuf, []byte("_client_name"))
+		connectAttrsBuf = appendLengthEncodedString(connectAttrsBuf, []byte("go-mysql-driver"))
+
+		for k, v := range mc.cfg.ConnectAttrs {
+			if k == "_client_name" {
+				// do not allow overwriting reserved values
```

Comment:
The original reasoning is unclear to me. Mainly because I based this Pull Request on @Vasfed work from #737.
However, I think it is a reasonable feature to be able to overwrite the `_client_name`.
Hence, I changed it in https://github.com/go-sql-driver/mysql/pull/1241/commits/b5117fc25c8379ff83f22144646ff94aee0cfc16

@Vasfed If you think otherwise, please let us know what you think.

Reply:
Please follow what official connectors do.

Reply:
`_client_name` should be the driver name.
User should use name starting without underscore for user attributes.
For example, `program_name` is convention to store application name.

> https://dev.mysql.com/doc/refman/8.0/en/performance-schema-connection-attribute-tables.html
> Many MySQL client programs set a program_name attribute with a value equal to the client name. For example, mysqladmin and mysqldump set program_name to mysqladmin and mysqldump, respectively. MySQL Shell sets program_name to mysqlsh.

Reply:
@methane From what I can see at the official connectors (see [Search result for "_client_name" in MySQL org](https://github.com/search?q=org%3Amysql+%22_client_name%22&type=code)), they set the default value of the driver's name as default (like we do in the PR), but also allow an overwrite.

Would you agree with this?
If not, what is your suggestion here?

Reply:
> From what I can see at the official connectors (see [Search result for "_client_name" in MySQL org](https://github.com/search?q=org%3Amysql+%22_client_name%22&type=code)), they set the default value of the driver's name as default (like we do in the PR), but also allow an overwrite.

Would you elaborate? I can not confirm what you seen.

As far as I quick looked the search results, official connectors do:

* Allow arbitrary attribute names, but they are overwritten by default value:
   * Python: https://github.com/mysql/mysql-connector-python/blob/c4729e84a6849fe584034663af39c0ccde92fe44/lib/mysql/connector/connection.py#L152
   * Java: https://github.com/mysql/mysql-connector-j/blob/e920b979015ae7117d60d72bcc8f077a839cd791/src/main/protocol-impl/java/com/mysql/cj/protocol/a/NativeAuthenticationProvider.java#L525-L546
   * C++: https://github.com/mysql/mysql-server/blob/3290a66c89eb1625a7058e0ef732432b6952b435/sql-common/client.cc#L5843-L5881

* Raise an Exception when user try to use attribute name starting with `_`:
   * C#: https://github.com/mysql/mysql-connector-net/blob/502d718bed8ca9cf81a3a0397574f24ec41b25ba/MySQL.Data/src/X/Sessions/XInternalSession.cs#L450-L469

So no official connectors allow setting user value for `_client_name`, `_os`, etc.

Reply:
@methane Thanks for the details.
What would be your recommended way?
I see multiple options:

1. Leave it as is (would not follow what official connectors do)
2. Skip/Ignore every attribute starting with a `_`
3. Let the user set arbitrary attribute names, but they are overwritten by default value

Happy to implement the preferred way.

On another note, I see that the following fields are not set (only `_client_name`):
- "_os"
- "_pid"
- "_platform"

Whats your opinion on setting these fields as well?

Reply:
> What would be your recommended way?

I prefer returning an error for name starting with `_`, because allow it later is backward compatible.

> On another note, I see that the following fields are not set (only `_client_name`):

I think `_pid` is useful. I don't think other attributes are useful, but OK to add them if we can get them easily.